be/src/tools/builder_helper.cpp

```diff
@@ -0,0 +1,295 @@
+
+    std::ifstream f(path, std::ios::in | std::ios::binary);
+    const auto sz = std::filesystem::file_size(path);
+    std::string result(sz, '\0');
+    f.read(result.data(), sz);
+
+    return result;
+}
+
+void BuilderHelper::build() {
+    // load tablet
+    std::string buf;
+    if (_isHDFS) {
+        FileSystemProperties system_properties;
+        LOG(INFO) << "read meta from HDFS:" << _meta_file;
+        auto* dfs = hdfsFileSystem.get();
+        bool exist = false;
+        auto st = dfs->exists(_meta_file, &exist);
+        if (!st.ok() || !exist) {
+            LOG(FATAL) << "file not exist:" /*<< dfs->GetLastError() */ << _meta_file;
+        }
+
+        size_t size = 0;
+        st = dfs->file_size(_meta_file, &size);
+        if (!st.ok() || size <= 0)
+            LOG(FATAL) << "meta file size abnormal.."; //<< dfs->GetLastError();
+        IOContext io_ctx;
+        io::FileReaderSPtr read_ptr = nullptr;
+        st = dfs->open_file(_meta_file, &read_ptr, &io_ctx);
+        if (!st.ok() || !read_ptr)
+            LOG(FATAL) << "cannot open file:" /*<< dfs->GetLastError()*/ << _meta_file;
+
+        buf.resize(size);
+
+        size_t read_size = 0;
+        read_ptr->read_at(0, Slice(buf.data(), size), io_ctx, &read_size);
+        if (read_size < size) {
+            LOG(FATAL) << "read meta file size abnormal.."; //<< dfs->GetLastError();
+        }
+
+        read_ptr->close();
+    } else {
+        buf = read_local_file(_meta_file);
+    }
+
+    FileHeader<TabletMetaPB> file_header;
+    buf = buf.substr(file_header.size());
+    // init tablet
+    TabletMeta* tablet_meta = new TabletMeta();
+    Status status = tablet_meta->deserialize(buf);
+    if (!status.ok()) {
+        LOG(FATAL) << "fail to load tablet meta :" << status.to_string();
+        return;
+    }
+
+    LOG(INFO) << "table id:" << tablet_meta->table_id() << " tablet id:" << tablet_meta->tablet_id()
+              << " shard id:" << tablet_meta->shard_id();
+
+    auto data_dir = StorageEngine::instance()->get_store(_build_dir);
+    TabletMetaSharedPtr tablet_meta_ptr(tablet_meta);
+    TabletSharedPtr new_tablet = doris::Tablet::create_tablet_from_meta(tablet_meta_ptr, data_dir);
+    status = StorageEngine::instance()->tablet_manager()->_add_tablet_unlocked(
+            tablet_meta->tablet_id(), new_tablet, false, true);
+
+    if (!status.ok()) {
+        LOG(FATAL) << "fail to add tablet to storage :" << status.to_string();
+        return;
+    }
+
+    std::vector<std::string> files;
+    if (_isHDFS) {
+        auto* dfs = hdfsFileSystem.get();
+        std::vector<io::Path> names;
+        dfs->list(_data_path, &names);
+        for (const auto& name : names) {
+            std::string filename = name.filename().c_str();
+            if (filename.substr(filename.size() - _file_type.size()) == _file_type) {
+                LOG(INFO) << "get file:" << name;
+                files.emplace_back(filename);
+            }
```

Comment:
suggest to add some warning log when the file type are not matched both in hdfs and local dir.